### PR TITLE
[SPARK-29042][Core] Sampling-based RDD with unordered input should be INDETERMINATE

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

We already have found and fixed the correctness issue before when RDD output is INDETERMINATE. One missing part is sampling-based RDD. This kind of RDDs is order sensitive to its input. A sampling-based RDD with unordered input, should be INDETERMINATE.

### Why are the changes needed?

A sampling-based RDD with unordered input is just like MapPartitionsRDD with isOrderSensitive parameter as true. The RDD output can be different after a rerun. 

It is a problem in ML applications.

In ML, sample is used to prepare training data. ML algorithm fits the model based on the sampled data. If rerun tasks of sample produce different output during model fitting, ML results will be unreliable and also buggy.

Each sample is random output, but once you sampled, the output should be determinate.

### Does this PR introduce any user-facing change?

Previously, a sampling-based RDD can possibly come with different output after a rerun.
After this patch, sampling-based RDD is INDETERMINATE. For an INDETERMINATE map stage, currently Spark scheduler will re-try all the tasks of the failed stage.

### How was this patch tested?

Added test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @felixcheung @cloud-fan

**Comment by HyukjinKwon** [comment]
@viirya, looks like technically it introduces a behaviour change ("Does this PR introduce any user-facing change?") assuming it affects determinism after a rerun given the description.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 1]
```diff
@@ -870,6 +870,29 @@ abstract class RDD[T: ClassTag](
...
   }
 
+  /**
+   * Return a new RDD by applying a function to each partition of this RDD, while tracking the index
+   * of the original partition.
+   *
+   * `preservesPartitioning` indicates whether the input function preserves the partitioner, which
+   * should be `false` unless this is a pair RDD and the input function doesn't modify the keys.
+   *
+   * `isOrderSensitive` indicates whether the function is order-sensitive. If it's order
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: no big deal but I usually avoid abbreviation in the doc though.

Comment (viirya) [reply core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Ok. Fixed.

**Comment by viirya** [comment]
@HyukjinKwon Thanks! I updated the PR description.

**Comment by viirya** [comment]
retest this please

**Review by felixcheung — APPROVED** [review]
LGTM. what's our policy now on correctness change?

**Comment by cloud-fan** [comment]
Do we have any queries return wrong result because of it?

for round-robin partitioner, it has an expectation that it should return the same output when rerun, otherwise we need to rerun the entire stage. This is for the correctness of `repartition`.

However, I don't think sample has the same problem. End-users would expect sample to return random output, so it doesn't matter if Spark returns different output when rerun tasks of sample.

**Comment by viirya** [comment]
I will go to merge this later if no other comments. We can decide to expose mapPartitionsWithIndex later if we want.

**Comment by viirya** [comment]
Merged to master. Thanks!

**Comment by gatorsmile** [comment]
@viirya Could you backport this to 2.4?

**Comment by viirya** [comment]
@gatorsmile Ok. Will do backport.

**Pull request closed without merging**